common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,381 @@
+//     the deletion when the store is next started.
+const metadataTableID uint32 = math.MaxUint32
+
+// The table ID reserved for the namespace table. Keys in the namespace table are table IDs (uint32)
+// and values are table names (string).
+const namespaceTableID uint32 = math.MaxUint32 - 1
+
+// This key is used to store the schema version in the metadata table.
+const metadataSchemaVersionKey = "schema_version"
+
+// The metadataDeletionKey holds the value of the table name that is being deleted if one is currently being deleted.
+// This is required for atomic deletion of tables.
+const metadataDeletionKey = "deletion"
+
+// The current schema version of the metadata table.
+const currentSchemaVersion uint64 = 0
+
+// StoreType describes the underlying store implementation.
+type StoreType int
+
+const (
+	// LevelDB is a LevelDB-backed store.
+	LevelDB StoreType = iota
+	// MapStore is an in-memory store. This store does not preserve data across restarts.
+	MapStore
+)
+
+// Start creates a new TableStore instance of the given type. The store will be created at the given path.
+// It can be used to instantiate a new store or to load an existing store.
+// This method will set up a table for each table name provided, and will drop all tables not in the list.
+// Dropping a table is irreversible and will delete all data in the table, so be very careful not to call
+// this method with table names omitted by mistake.
+func (t StoreType) Start(logger logging.Logger, path string, tables ...string) (kvstore.TableStore, error) {
+
+	var base kvstore.Store
+	var err error
+
+	switch t {
+	case LevelDB:
+		base, err = leveldb.NewStore(logger, path)
+		if err != nil {
+			return nil, fmt.Errorf("error creating LevelDB store: %w", err)
+		}
+	case MapStore:
+		base = mapstore.NewStore()
+	default:
+		return nil, fmt.Errorf("unknown store type: %d", t)
+	}
+
+	return start(logger, base, tables...)
+}
+
+// Future work: if we ever decide to permit third parties to provide custom store implementations not in this module,
+// we will need to make this method public.
+
+// start creates a new TableStore instance with the given base store and table names.
+func start(logger logging.Logger, base kvstore.Store, tables ...string) (kvstore.TableStore, error) {
+	metadataTable := newTableView(base, "metadata", metadataTableID)
+	namespaceTable := newTableView(base, "namespace", namespaceTableID)
+
+	err := validateSchema(metadataTable)
+	if err != nil {
+		return nil, fmt.Errorf("error validating schema: %w", err)
+	}
+
+	tableIDMap, err := loadNamespaceTable(namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error loading namespace table: %w", err)
+	}
+
+	err = handleIncompleteDeletion(logger, base, metadataTable, namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error handling incomplete deletion: %w", err)
+	}
+
+	err = addAndRemoveTables(base, metadataTable, namespaceTable, tableIDMap, tables)
```

Comment:
I'd prefer not deleting tables synchronously. After the database running for a while, the deletion can take quite a while. It'd be better to:
- Do lazy deletion: write the table to be deleted to the deletion table
- Run a thread in the background to perform the deletion - and cap the resource it can use

Reply:
We discussed this, repeating it here for the sake of others. This PR is already fairly large, and so we've decided to implement async table deletion as a follow up PR.